### PR TITLE
[WIP] NuCypher ConnectNative Firefox Extension 

### PR DESCRIPTION
**Pull request by vepkenez**
* extracted from Demo and rebased on master
* An install script!
* A readme!
* Status check to tell you if it's installed correctly

requires #1555

### CODE REVIEW

**Comment by tuxxy** [comment]
What's the status of this PR? What do we need to get it merged?

**Comment by michwill** [comment]
Looks like putting a required `value` argument breaks (currently federated) character tests, as well as RPC API applied to federated network.
I've started working on a PR which allows optional arguments in the specs https://github.com/nucypher/nucypher/pull/1366. Perhaps, that's the way to solve. But since this problem was encountered and worked around here, I need to reference!

**Comment by michwill** [comment]
Well, actually I'm wrong. It works with RPC in general, just some bugs with web RPC in tests

**Comment by michwill** [comment]
Ok, looks like tests for optional arguments work properly in https://github.com/nucypher/nucypher/pull/1366. I think, if this PR will be rebased on top of that (will need to replace the specs), the character tests will pass

**Comment by KPrasch** [comment]
Awaiting PR #1366

**Comment by KPrasch** [comment]
Merged - Rebase needed.

**Comment by michwill** [comment]
I wonder if rebase will fix the tests, or some touches are required

**Comment by vepkenez** [comment]
rebased. lets see.

**Comment by vepkenez** [comment]
also what is "birdcage"?

**Comment by michwill** [comment]
Character tests pass!

**Comment by vepkenez** [comment]
I think this should probably be based over master?

**Comment by michwill** [comment]
The necessary change with optional arguments is already in master, yes

**Comment by vepkenez** [comment]
Yeah, that reduced the "files changed from ~72 to 16

**Comment by vepkenez** [comment]
Ok this PR needs some tests.  Some things aren't working for some reason so I'm going to put a [WIP] on it and  finish writing tests around --options.

**Comment by KPrasch** [comment]
Installation scripts work, however it looks like the `--options` flag didn't make it into the codebase, leading to `result: Usage: nucypher-cli status [OPTIONS] ACTION Try "nucypher-cli status --help" for help. Error: no such option: --options `

**Comment by vepkenez** [comment]
@KPrasch yes I think the major CLI refactor which happened after this was all done is necessitating some re-working of this.

File: apps/connectnative/install/firefox.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        'Library/Application Support/Mozilla/ManagedStorage/',
+        'Library/Application Support/Mozilla/PKCS11Modules/',
+    ]
+}
+
+def install():
+    this_os = platform.system()
+    home = expanduser("~")
+
+    stdio_cli = shutil.which('nucypher-stdio-receiver')
```
Comment (tuxxy) [inline apps/connectnative/install/firefox.py]: This is neat.

File: nucypher/cli/characters/alice.py [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
 from nucypher.cli.config import nucypher_click_config
 from nucypher.cli.types import NETWORK_PORT, EXISTING_READABLE_FILE, EIP55_CHECKSUM_ADDRESS
 from nucypher.config.characters import AliceConfiguration
 from nucypher.config.keyring import NucypherKeyring
 
 
+
 @click.command()
 @click.argument('action')
+@click.option('--options', help="Export JSON type schema", is_flag=True)
```
Comment (tuxxy) [inline nucypher/cli/characters/alice.py]: The name of this option and the help don't really help me understand either here.

File: nucypher/cli/characters/enrico.py [hunk 3]
```diff
@@ -5,16 +5,19 @@
 from nucypher.characters.lawful import Enrico
 from nucypher.cli.config import nucypher_click_config
 from nucypher.cli.types import NETWORK_PORT
+from nucypher.cli.actions import echo_schema
+from nucypher.characters.control.specifications import EnricoSpecification
 
 
 @click.command()
 @click.argument('action')
+@click.option('--options', help="Export JSON type schema", is_flag=True)
```
Comment (tuxxy) [inline nucypher/cli/characters/enrico.py]: Same critique here, too.

**Comment by KPrasch** [comment]
Continuing the above dialog - How do we want to proceed with this PR?

**Comment by michwill** [comment]
There are some conflicts but... Doesn't seem to be breaking anything, so why not merge if those are fixed?

**Comment by vepkenez** [comment]
I think this PR will need a bit of work to adapt it to some of the CLI changes of late, but after that it will work better and will cover more functionality than it originally did when this was opened.

**Pull request closed without merging**